### PR TITLE
repeat create session on fail

### LINKED ISSUE #184: Repeate create camera session on fail
Currently if create camera session fails we just throw an exception. So user will have a crash.
We should change logic to repeat creation for 3 times. If it doesn't help - close camera.
That logic may be changed/improved in [1512](https://app.zenhub.com/workspaces/vision-sdk-5ca52e6c027e3e7234ac7512/issues/mapbox/mapbox-vision/1512)

### PR DESCRIPTION
**Pull request by korshaknn**
closes #184

### CODE REVIEW

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 1]
```diff
@@ -167,7 +170,14 @@ class Camera2VideoSourceImpl(
                         }
 
                         override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
-                            throw IllegalStateException("Failed to configure Camera ${cameraCaptureSession.device.id}!")
+                            if (attemptNumber <= attemptMax) {
```
Comment (alexander-kulikovskii) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: will _attemptMax_ be converted to _Long_ each time?

Comment (korshaknn) [reply MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: ptal after refactoring

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 2]
```diff
@@ -37,6 +37,8 @@ class Camera2VideoSourceImpl(
 
     private var sensorOrientation: Int = 0
 
+    private var attemptNumber = 1
```
Comment (RingerJK) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]:
```suggestion
    private var attemptNumber = 0
```

Comment (RingerJK) [reply MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: and in `createCaptureSession` do `attemptNumber++`

Comment (RingerJK) [reply MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: in `onConfigured `  -> `attemptNumber = 0`

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 3]
```diff
@@ -167,7 +169,14 @@ class Camera2VideoSourceImpl(
...
 
                         override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
-                            throw IllegalStateException("Failed to configure Camera ${cameraCaptureSession.device.id}!")
+                            if (attemptNumber <= CREATE_SESSION_MAX_ATTEMPTS) {
+                                backgroundThreadHandler.handler.postDelayed({createCaptureSession()}, TimeUnit.SECONDS.toMillis(attemptNumber * 2L))  // delay 2 - 4 - 6 seconds
+                                attemptNumber++
+                            } else {
+                                cameraDevice?.close()
+                                cameraDevice = null
+                                cameraOpenCloseLock.release()
```
Comment (RingerJK) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: `attemptNumber  = 0`

**Comment by korshaknn** [comment]
@RingerJK  fixed, ptal

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 4]
```diff
@@ -167,7 +171,14 @@ class Camera2VideoSourceImpl(
                         }
 
                         override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
-                            throw IllegalStateException("Failed to configure Camera ${cameraCaptureSession.device.id}!")
+                            if (attemptNumber <= CREATE_SESSION_MAX_ATTEMPTS) {
+                                backgroundThreadHandler.handler.postDelayed({createCaptureSession()}, TimeUnit.SECONDS.toMillis(attemptNumber * 2L))  // delay 2 - 4 - 6 seconds
+                            } else {
+                                attemptNumber = 0
+                                cameraDevice?.close()
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: As it's the same 3 calls for this - extract to private fun?

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 5]
```diff
@@ -37,6 +37,8 @@ class Camera2VideoSourceImpl(
 
     private var sensorOrientation: Int = 0
 
+    private var attemptNumber = 0
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: `configureAttemptNumber`?

**Comment by korshaknn** [comment]
@yunikkk  done, ptal

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt [hunk 6]
```diff
@@ -178,6 +182,17 @@ class Camera2VideoSourceImpl(
         }
     }
 
+    private fun repeatConfigureAttempt() {
+        if (configureAttemptNumber <= CREATE_SESSION_MAX_ATTEMPTS) {
+            backgroundThreadHandler.handler.postDelayed({createCaptureSession()}, TimeUnit.SECONDS.toMillis(configureAttemptNumber * 2L))  // delay 2 - 4 - 6 seconds
+        } else {
+            configureAttemptNumber = 0
+            cameraDevice?.close()
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: I meant lines 
```
cameraDevice?.close()
cameraDevice = null
cameraOpenCloseLock.release()
```
that are used in two other places, too

Comment (korshaknn) [reply MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/Camera2VideoSourceImpl.kt]: fixed

**Pull request merged**